src/persist-txn/src/txns.rs

```diff
@@ -210,7 +215,7 @@ where
         &mut self,
         register_ts: T,
         data_writes: impl IntoIterator<Item = WriteHandle<K, V, T, D>>,
-    ) -> Result<(), T> {
+    ) -> Result<TxnApply<T>, T> {
```

Comment:
IMO we should pull the `apply` call into the register impl. I can't imagine why a caller would want to split up the two steps, and if we come across one we can switch back to what you have here. I like the simplicity as well as the increased invariants (i.e. the physical upper of the data shard will be past register_ts when this returns)

(I'd also return the `Tidy` resulting from the apply call so that the storage controller can save doing the apply work twice, but I'd probably just ignore doing anything with it in existing tests)

Reply:
One concern is that executing the apply in this method can make this method have surprising latency. In some scenarios registering a new shard will end up applying the results of a bunch of other transactions.

Reply:
Yup, agreed, but that's already an issue with apply in general. I think to resolve [this TODO](https://github.com/MaterializeInc/materialize/blob/f30494b8fc53a9e81cba6aa90c1431d21492b486/src/persist-txn/src/txns.rs#L138-L141) about weird placeholder schemas, I'll have to make apply more targeted (i.e. a specific set of data shards), in which case this method would ignore work for other data shards. So, I'm personally not too worried about this